L1Trigger/TrackFindingTracklet/interface/KFin.h

```diff
@@ -0,0 +1,136 @@
+    void consume(const tt::StreamsTrack& streamsTrack, const tt::StreamsStub& streamsStub);
+    // fill output products
+    void produce(tt::StreamsStub& accpetedStubs,
+                 tt::StreamsTrack& acceptedTracks,
+                 tt::StreamsStub& lostStubs,
+                 tt::StreamsTrack& lostTracks);
+
+  private:
+    // truncates double precision of val into base precision
+    double digi(double val, double base) const { return (floor(val / base + 1.e-12) + .5) * base; }
+    //double digis(double val, double base) const { return (floor(val / base + .5) + .5) * base; }
+    // basetransformation of val from baseLow into baseHigh using widthMultiplier bit multiplication
+    double redigi(double val, double baseLow, double baseHigh, int widthMultiplier) const;
+    struct Stub {
+      Stub(const TTStubRef& ttStubRef, int layer, double r, double phi, double z, bool psTilt)
+          : valid_(true), ttStubRef_(ttStubRef), layer_(layer), r_(r), phi_(phi), z_(z), psTilt_(psTilt) {}
+      bool valid_;
+      TTStubRef ttStubRef_;
+      int layer_;
+      double r_;
+      double phi_;
+      double z_;
+      double dPhi_;
+      double dZ_;
+      bool psTilt_;
+    };
+    struct Track {
+      static constexpr int max_ = 8;
+      Track() { stubs_.reserve(max_); }
+      Track(const TTTrackRef& ttTrackRef,
+            bool valid,
+            double inv2R,
+            double phiT,
+            double cot,
+            double zT,
+            const std::vector<Stub*>& stubs)
+          : ttTrackRef_(ttTrackRef),
+            valid_(valid),
+            sector_(-1),
+            inv2R_(inv2R),
+            phiT_(phiT),
+            cot_(cot),
+            zT_(zT),
+            stubs_(stubs) {}
+      TTTrackRef ttTrackRef_;
+      bool valid_;
+      TTBV maybe_;
+      int sector_;
+      double inv2R_;
+      double phiT_;
+      double cot_;
+      double zT_;
+      std::vector<Stub*> stubs_;
+    };
+
+    // true if truncation is enbaled
+    bool enableTruncation_;
+    // provides run-time constants
+    const tt::Setup* setup_;
+    // provides dataformats
+    const trackerTFP::DataFormats* dataFormats_;
+    // helper class to encode layer
+    const trackerTFP::LayerEncoding* layerEncoding_;
+    // helper class to assign tracks to channel
+    const ChannelAssignment* channelAssignment_;
+    // provides tracklet constants
+    const Settings* settings_;
+    // processing region (0 - 8) aka processing phi nonant
+    const int region_;
+    // storage of input tracks
+    std::vector<Track> tracks_;
+    // storage of input stubs
+    std::vector<Stub> stubs_;
+    // h/w liked organized pointer to input tracks
+    std::vector<std::vector<Track*>> input_;
+    // unified tracklet bases
+    double baseUinv2R_;
+    double baseUphiT_;
+    double baseUcot_;
+    double baseUzT_;
+    double baseUr_;
+    double baseUphi_;
+    double baseUz_;
+    // low precise bases KF is using
```

Comment:
I don't understand the comment "low precise bases". I believe KF uses finer granularity than tracklet, so this comment should say "higher granularity bases"?

Reply:
KF expects much lower bases for the input tracks as tracklet delivers, therefore I call them "low".

Reply:
I don't think this is good English. The base is always precisely known, so the word "precise" can be removed here. 
From a google search, I can't find any use of "digitisation base" in the context of FPGAs. Is this a recognised term? I therefore suspect people won't understand what you mean by "low base".

Reply:
done